## Pearson Product Moment Correlation Coefficient


A correlation is the degree of association between two variables. One of my favorite books on this topic is: 

<img src="http://zacharski.org/files/courses/cs419/mg_statistics_big.png" width="250" />

and they illustrate it by looking at
## Ladies expenditures on clothes and makeup

<img src="http://zacharski.org/files/courses/cs419/correlation1.png" width="650" />
<img src="http://zacharski.org/files/courses/cs419/corr2.png" width="650" />

So let's go ahead and create that data in pandas and show the table:

In [1]:
import pandas as pd
from pandas import DataFrame
makeup =  [3000, 5000, 12000, 2000, 7000,  15000,  5000,  6000, 8000,  10000]
clothes = [7000, 8000, 25000, 5000, 12000, 30000, 10000, 15000, 20000, 18000]
ladies = ['Ms A','Ms B','Ms C','Ms D','Ms E','Ms F','Ms G','Ms H','Ms I','Ms J',]
monthly = DataFrame({'makeup': makeup, 'clothes': clothes}, index= ladies)
monthly

clothes  makeup
Ms A     7000    3000
Ms B     8000    5000
Ms C    25000   12000
Ms D     5000    2000
Ms E    12000    7000
Ms F    30000   15000
Ms G    10000    5000
Ms H    15000    6000
Ms I    20000    8000
Ms J    18000   10000

and let's show the scatterplot

In [18]:
from bokeh.charts import Scatter, output_file, show
from bokeh.io import push_notebook, show, output_notebook
output_notebook()
x = Scatter(monthly, x='makeup', y='clothes', title="Montly Expenditures on Makeup and Clothes",
            xlabel="Money spent on makeup", ylabel="Money spent on clothes")

output_file("stuff.html")

show(x)

Loading BokehJS ...

INFO:bokeh.core.state:Session output file 'stuff.html' already exists, will be overwritten.


When the data points are close to a straight line going up, we say that there is a positive correlation between the two variables. Let's look at a few more examples:

## Weight and calories in 1-3 yr/old children


In [34]:
weight = [7.7, 7.8, 8.6, 8.5, 8.6, 9, 10.1, 11.5, 11, 10.2, 11.9, 10.4, 9.3, 9.1, 8.5, 11]
calories = [360, 400, 500, 370, 525, 800, 900, 1200, 1000, 1400, 1600, 850, 575, 425, 950, 800]
kids = DataFrame({'weight': weight, 'calories': calories})
kids

calories  weight
0        360     7.7
1        400     7.8
2        500     8.6
3        370     8.5
4        525     8.6
5        800     9.0
6        900    10.1
7       1200    11.5
8       1000    11.0
9       1400    10.2
10      1600    11.9
11       850    10.4
12       575     9.3
13       425     9.1
14       950     8.5
15       800    11.0

In [35]:
x = Scatter(kids, x='weight', y='calories', title="Weight and calories in 1-3 yr.old children",
            xlabel="weight (kg)", ylabel="weekly calories")

output_file("stuff.html")

show(x)

INFO:bokeh.core.state:Session output file 'stuff.html' already exists, will be overwritten.


## The stronger the correlation the closer to a straight line:


In [38]:
midterm = [65, 70, 72, 75, 78, 78, 80, 83, 87, 87, 90, 90, 92, 93, 95, 100]
final =   [72, 68, 75, 76, 79, 73, 79, 85, 89, 87, 92, 93, 94, 91, 97, 98]
grades = DataFrame({'midterm': midterm, 'final': final})
grades

final  midterm
0      72       65
1      68       70
2      75       72
3      76       75
4      79       78
5      73       78
6      79       80
7      85       83
8      89       87
9      87       87
10     92       90
11     93       90
12     94       92
13     91       93
14     97       95
15     98      100

In [44]:

scatterGrades = Scatter(grades, x='midterm', y='final', title="Comparison of Midterm and Final Grades",
            xlabel="Midterm grade", ylabel="Final grade", color='navy')

output_file("grades.html")

show(scatterGrades)

INFO:bokeh.core.state:Session output file 'grades.html' already exists, will be overwritten.


## perfect positive correlation

Here I am comparing the heights in inches of some NHL players with their heights in cm.

In [46]:
inches =[68, 73, 69,72,71,77]
cm = [173, 185, 175, 183, 180, 196]
nhlHeights = DataFrame({'heightInches': inches, 'heightCM': cm})
nhlHeights

heightCM  heightInches
0       173            68
1       185            73
2       175            69
3       183            72
4       180            71
5       196            77

In [50]:
heights = Scatter(nhlHeights, x='heightInches', y='heightCM', title="Comparison of Height in Inches and Height in CM",
            xlabel="Height in Inches", ylabel="Height in centimeters", color='indigo')

output_file("heights.html")

show(heights)

INFO:bokeh.core.state:Session output file 'heights.html' already exists, will be overwritten.


In [3]:
medicineHat = pd.read_csv('https://raw.githubusercontent.com/zacharski/machine-learning/master/data/medicineHatTigers.csv')
medicineHat

Name Pos. L/R  Height  Weight  Birthday          Hometown
0       Brennan Bosch    C   R      68     173  02/14/88  Martensville, SK
1        Scott Wasden    C   R      73     188  01/04/88      Westbank, BC
2        Colton Grant   LW   L      69     177  03/20/89      Standard, AB
3         Darren Helm   LW   L      72     182  01/21/87   St. Andrews, MB
4       Derek Dorsett   RW   L      71     178  12/20/86    Kindersley, SK
5          Daine Todd    C   R      70     173  01/10/87      Red Deer, AB
6       Tyler Swystun   RW   R      71     185  01/15/88      Cochrane, AB
7          Matt Lowry    C   R      72     186  03/02/88       Neepawa, MB
8   Kevin Undershute    LW   L      72     178  04/12/87  Medicine Hat, AB
9        Jerrid Sauer   RW   R      71     196  09/12/87  Medicine Hat, AB
10        Tyler Ennis    C   L      69     160  10/06/89      Edmonton, AB
11    Jordan Hickmott    C   R      72     183  04/11/90        Mission BC
12       Jakub Rumpel   RW   R      68     166  01/27/87  Hrnciarovce, SLO
13    Bretton Cameron    C   R      71     168  01/26/89      Didsbury, AB
14      Chris Stevens   LW   L      70     197  08/20/86  Dawson Creek, BC
15       Gord Baldwin    D   L      77     205  03/01/87      Winnipeg, MB
16     David Schlemko    D   L      73     195  05/07/87      Edmonton, AB
17       Trever Glass    D   L      72     190  01/22/88      Cochrane, AB
18       Kris Russell    D   L      70     177  05/02/87      Caroline, AB
19     Michael Sauer     D   R      73     205  08/07/87       Sartell, MN
20     Mark Isherwood    D   R      72     183  01/31/89    Abbotsford, BC
21      Shayne Brown     D   L      73     198  02/20/89   Stony Plain, AB
22    Jordan Bendfeld    D   R      75     230  02/09/88         Ledue, AB
23       Ryan Holfeld    G   L      71     166  06/29/89         LeRoy, SK
24       Matt Keetley    G   R      74     189  04/27/86  Medicine Hat, AB

In [57]:
medicineHat['hometownLength'] = medicineHat['Hometown'].str.len()
medicineHat

Name Pos. L/R  Height  Weight  Birthday          Hometown  \
0       Brennan Bosch    C   R      68     173  02/14/88  Martensville, SK   
1        Scott Wasden    C   R      73     188  01/04/88      Westbank, BC   
2        Colton Grant   LW   L      69     177  03/20/89      Standard, AB   
3         Darren Helm   LW   L      72     182  01/21/87   St. Andrews, MB   
4       Derek Dorsett   RW   L      71     178  12/20/86    Kindersley, SK   
5          Daine Todd    C   R      70     173  01/10/87      Red Deer, AB   
6       Tyler Swystun   RW   R      71     185  01/15/88      Cochrane, AB   
7          Matt Lowry    C   R      72     186  03/02/88       Neepawa, MB   
8   Kevin Undershute    LW   L      72     178  04/12/87  Medicine Hat, AB   
9        Jerrid Sauer   RW   R      71     196  09/12/87  Medicine Hat, AB   
10        Tyler Ennis    C   L      69     160  10/06/89      Edmonton, AB   
11    Jordan Hickmott    C   R      72     183  04/11/90        Mission BC   
12       Jakub Rumpel   RW   R      68     166  01/27/87  Hrnciarovce, SLO   
13    Bretton Cameron    C   R      71     168  01/26/89      Didsbury, AB   
14      Chris Stevens   LW   L      70     197  08/20/86  Dawson Creek, BC   
15       Gord Baldwin    D   L      77     205  03/01/87      Winnipeg, MB   
16     David Schlemko    D   L      73     195  05/07/87      Edmonton, AB   
17       Trever Glass    D   L      72     190  01/22/88      Cochrane, AB   
18       Kris Russell    D   L      70     177  05/02/87      Caroline, AB   
19     Michael Sauer     D   R      73     205  08/07/87       Sartell, MN   
20     Mark Isherwood    D   R      72     183  01/31/89    Abbotsford, BC   
21      Shayne Brown     D   L      73     198  02/20/89   Stony Plain, AB   
22    Jordan Bendfeld    D   R      75     230  02/09/88         Ledue, AB   
23       Ryan Holfeld    G   L      71     166  06/29/89         LeRoy, SK   
24       Matt Keetley    G   R      74     189  04/27/86  Medicine Hat, AB   

    hometownLength  
0               16  
1               12  
2               12  
3               15  
4               14  
5               12  
6               12  
7               11  
8               16  
9               16  
10              12  
11              10  
12              16  
13              12  
14              16  
15              12  
16              12  
17              12  
18              12  
19              11  
20              14  
21              15  
22               9  
23               9  
24              16

In [59]:
medicineHat[['Height','hometownLength']]

Height  hometownLength
0       68              16
1       73              12
2       69              12
3       72              15
4       71              14
5       70              12
6       71              12
7       72              11
8       72              16
9       71              16
10      69              12
11      72              10
12      68              16
13      71              12
14      70              16
15      77              12
16      73              12
17      72              12
18      70              12
19      73              11
20      72              14
21      73              15
22      75               9
23      71               9
24      74              16

In [66]:
hometownHeights = Scatter(medicineHat[['Height','hometownLength']], x='hometownLength', y='Height', title="Correlation of the number of Letters in the Hometown to Height",
            xlabel="Height in Inches", ylabel="Height in centimeters", color='teal', marker='square')

output_file("hometown.html")

show(hometownHeights)

INFO:bokeh.core.state:Session output file 'hometown.html' already exists, will be overwritten.


## negative correlation has a line going downhill

#### the relationship of hair length to a person's height

In [69]:
height =[62, 64, 65, 68, 69, 70, 67, 65, 72, 73, 74]
hairLength = [7, 10, 6, 4, 5, 4, 5, 8, 1, 1, 3]
cm = [173, 185, 175, 183, 180, 196]
people = DataFrame({'height': height, 'hairLength': hairLength})
hair = Scatter(people, x='height', y='hairLength', title="Comparison of an Individual's Height to the Length of their Hair",
            xlabel="Height in Inches", ylabel="Hair Length in Inches", color='indigo')

output_file("hair.html")

show(hair)

The taller someone is, the shorter their hair.


####  The Effects of Alcohol on the Dexterity Score
#### a strong negative correlation

In [74]:
drinks = list(range(11))
dexterity = [96, 96, 90, 63, 65, 50, 47, 46, 18, 17, 9]
drinking = DataFrame({'drinks': drinks, 'dexterity': dexterity})
drinkChart = Scatter(drinking, x='drinks', y='dexterity', title="The effects of alcohol consumption on dexterity",
            xlabel="Number of Drinks", ylabel="Dexterity Score", color='maroon')

output_file("drink.html")

show(drinkChart)


# Numeric Represenstation of the Strength of the Correlation


## Pearson Correlation Coefficient

This ranges from -1 to 1.
1 is perfect positive correlation, -1 is perfect negative.

$$r=\frac{\sum_{i=1}^n(x_i - \bar{x})(y_i-\bar{y})}{\sqrt{\sum_{i=1}^n(x_i - \bar{x})}  \sqrt{\sum_{i=1}^n(y_i - \bar{y})}}$$

You should know how to compute the formula by hand. For help, checkout the textbook. In Pandas it is very easy to compute.

### Japanese ladies expensives on makeup and clothes
Let's go back to our first example.
First here is the data:

In [77]:
monthly

clothes  makeup
Ms A     7000    3000
Ms B     8000    5000
Ms C    25000   12000
Ms D     5000    2000
Ms E    12000    7000
Ms F    30000   15000
Ms G    10000    5000
Ms H    15000    6000
Ms I    20000    8000
Ms J    18000   10000

And our Scatter plot looked like this:

In [79]:
x = Scatter(monthly, x='makeup', y='clothes', title="Montly Expenditures on Makeup and Clothes",
            xlabel="Money spent on makeup", ylabel="Money spent on clothes")

output_file("stuff.html")

show(x)

INFO:bokeh.core.state:Session output file 'stuff.html' already exists, will be overwritten.


So that looks like a pretty strong positive correlation. To compute Pearson on this data we do:

In [80]:
monthly.corr()

clothes   makeup
clothes  1.00000  0.96802
makeup   0.96802  1.00000

There is the obvious: clothes expenditures are perfectly correlated with clothes expenditures. But the correlation between clothes and makeup is also very strong at 0.96802.

What about my idea that the number of characters in the name of someone's hometown is correlated with the height of that person:


In [81]:
medicineHat.corr()

Height    Weight  hometownLength
Height          1.000000  0.736502       -0.277811
Weight          0.736502  1.000000       -0.156846
hometownLength -0.277811 -0.156846        1.000000

The Pearson Correlation Coefficient for height to hometownLength is -0.2778 and that is pretty weak.

Finally, what about the number of drinks to dexterity:

In [83]:
drinking.corr()

dexterity    drinks
dexterity   1.000000 -0.979237
drinks     -0.979237  1.000000

So -0.97923 indicates a very strong negative correlation

<h2 style="color:red">You Try</h2>
<h3 style="color:red">1. Medicine Hat Tigers 10xp</h3>
<span style="color:red">We are interested in whether there is a relationship between the height and weight of the Medicine Hat Players.</span>

<h4 style="color:red">a. first show the scatter plot</h4>


In [4]:
# your code here

<h4 style="color:red">b. Now computer Pearson</h4>

In [6]:
# your code here


So what do you think? Is there a correlation between height and weight?

<h3 style="color:red">2. A School Music Program: A case study  15xp</h3>
<span style="color:red">We are interested in seeing whether a relationship exists between music performance
grades and students' individual Iowa Test of Basic Skills (ITBS) scores. Using the
 test set at</span>



In [7]:
musicFile = "https://raw.githubusercontent.com/zacharski/machine-learning/master/data/musicITBS.csv"

<span style="color:red">Calculate (using Pearson) the strength of the association between music and language,
music and science, and music and math. Are any of these associations strong?</span>



In [14]:
# Your code here


<span style="color:red">Please graph (using a scatter chart) the strongest association</span>



<h3 style="color:red">3. New Mexico School Districts 15xp</h3>
This requires a bit more work getting started.  The following file (http://zacharski.org/files/courses/cs419/DistrictGrades2016.xlsx) is a spreadsheet (**not** a CSV file). This is data from 2016 on New Mexico School Districts. This is also a bit of practice in data wrangling--getting the spreadsheet data in a form compatible with Pandas.
<span style="color:red">I want to know if there is a relationship between the number of schools in a school district and the quality (overall points) of the district. Show the correlation both visually and with a coefficient.</span>

<h3 style="color:red">4. Our Class 25xp </h3>
<span style="color:red">Using the Google spreadsheet of our class music rating, I would like to see the correlation matrix (I sort of made that term up) showing the Pearson Correlation for each pair of students in the class.
</span>

<h3 style="color:red">BONUS. Hacker Challenge: Nearest Neighbor   20-35xp</h3>
<span style="color:red">Can you write a function that returns the nearest neighbor of a student? (20xp)  

    def pearsonNearestNeighbor(adataframe, aStringRepresentingApersonsName, anyOtherParametersYouMightNeed):
</span>

Let me give you a sorting example that might prove useful. Let's say I have a DataFrame of students and their grades:

In [13]:
students = ['Ann', 'Berkley', 'Clara', 'Dora', 'Eiko']
grades = [93, 89, 96, 76, 91]
fg = DataFrame({'student': students, 'grade': grades})
finalGrades = fg.set_index('student')
finalGrades

grade
student       
Ann         93
Berkley     89
Clara       96
Dora        76
Eiko        91

Now let's say I want to sort by grade and also get the name of the student with the highest grade:

In [17]:

temp = finalGrades.sort_values('grade', ascending=False)
temp.index[0]

'Clara'

So Clara is the person with the highest grade.  Hopefully this will help you with the current problem. Googling for Pandas information is pretty fruitful.

<span style="color:red">Can you modify the function so it makes recommendations? (+15xp)  </span>